File: examples/scripts/hello-world.pre-rewrite.pl
#!/usr/bin/env perl
#
# hello-world.pre-rewrite.pl — the simplest possible --pre-rewrite-script.
#
# A pre-rewrite script is just a filter: it reads the raw `git fast-export`
# byte stream on stdin and writes a (possibly edited) stream to stdout, *before*
# `git filter-repo` parses it. The orchestrator runs it as:
#
#   git fast-export ... | <this script> | git filter-repo --stdin ...
#
# This example does ONE easy-to-follow thing: rewrite a single author/committer
# identity across all history — change "Old Name <old@example.com>" into
# "New Name <new@example.com>". Edit the two variables below for your own case.
#
# It still respects the two rules every pre-rewrite script must follow:
#   1. read/write raw bytes (binmode), never decoded text; and
#   2. copy `data <N>` payloads (blobs, commit messages) through untouched, so a
#      line *inside a file* that starts with "author " is never mistaken for an
#      identity line.
#
# For the real-world malformed-ident repair — and the full set of correctness
# guarantees — see repair-malformed-idents.pre-rewrite.pl in this folder.

use strict;
use warnings;

binmode STDIN;
binmode STDOUT;

# --- edit these two lines ----------------------------------------------------
my $OLD = 'Old Name <old@example.com>';
my $NEW = 'New Name <new@example.com>';
# -----------------------------------------------------------------------------

while (defined(my $line = <STDIN>)) {
    # `data <N>` introduces N raw bytes (a blob or a commit/tag message). Copy
    # them through verbatim so we never accidentally edit file contents.
    if ($line =~ /^data (\d+)\n\z/) {
        print $line;
        my $remaining = $1;
        while ($remaining > 0) {
            my $got = read(STDIN, my $chunk, $remaining);
            last unless $got;    # truncated stream; let downstream error out
            print $chunk;
            $remaining -= $got;
        }
        next;
    }

    # Only rewrite the identity on author / committer / tagger lines.
    if ($line =~ /^(?:author|committer|tagger) /) {
        $line =~ s/\Q$OLD\E/$NEW/;
    }

    print $line;
}

File: examples/scripts/hello-world.pre-rewrite.py
#!/usr/bin/env python3
#
# hello-world.pre-rewrite.py — the simplest possible --pre-rewrite-script.
#
# Reads the `git fast-export` stream on stdin, rewrites a single author/committer
# identity across all history, and writes the result to stdout. Edit the two
# variables below for your own case. Any language works — this one is Python;
# see repair-malformed-idents.pre-rewrite.pl for a real Perl example.

import re
import sys

# --- edit these two lines ----------------------------------------------------
OLD = b"Old Name <old@example.com>"
NEW = b"New Name <new@example.com>"
# -----------------------------------------------------------------------------

stdin = sys.stdin.buffer
stdout = sys.stdout.buffer

data_header = re.compile(rb"^data (\d+)\n\Z")
ident_line = re.compile(rb"^(?:author|committer|tagger) ")

while True:
    line = stdin.readline()
    if not line:
        break

    # Copy `data <N>` payloads (blobs, messages) through verbatim, byte for byte.
    m = data_header.match(line)
    if m:
        stdout.write(line)
        remaining = int(m.group(1))
        while remaining > 0:
            chunk = stdin.read(remaining)
            if not chunk:
                break  # truncated stream; let downstream error out
            stdout.write(chunk)
            remaining -= len(chunk)
        continue

    # Only rewrite the identity on author / committer / tagger lines.
    if ident_line.match(line):
        line = line.replace(OLD, NEW)

    stdout.write(line)
